src/ReactART.js

```diff
@@ -12,312 +12,71 @@
+const ReactFiberReconciler = require('react-dom/lib/ReactFiberReconciler');
 
-const emptyObject = require('fbjs/lib/emptyObject');
-const invariant = require('fbjs/lib/invariant');
+const { Component } = React;
 
-const assign = require('object-assign');
 const pooledTransform = new Transform();
 
-// Utilities
+const EVENT_TYPES = {
+  onClick: 'click',
+  onMouseMove: 'mousemove',
+  onMouseOver: 'mouseover',
+  onMouseOut: 'mouseout',
+  onMouseUp: 'mouseup',
+  onMouseDown: 'mousedown',
+};
+
+const TYPES = {
+  CLIPPING_RECTANGLE: 'ClippingRectangle',
+  GROUP: 'Group',
+  SHAPE: 'Shape',
+  TEXT: 'Text',
+};
+
+/** UTILITIES */
 
 function childrenAsString(children) {
   if (!children) {
     return '';
-  }
-  if (typeof children === 'string') {
+  } else if (typeof children === 'string') {
     return children;
-  }
-  if (children.length) {
+  } else if (children.length) {
     return children.join('\n');
-  }
-  return '';
-}
-
-function createComponent(name) {
-  const ReactARTComponent = function(element) {
-    this.node = null;
-    this.subscriptions = null;
-    this.listeners = null;
-    this._mountImage = null;
-    this._renderedChildren = null;
-    this.construct(element);
-  };
-  ReactARTComponent.displayName = name;
-  for (let i = 1, l = arguments.length; i < l; i++) {
-    assign(ReactARTComponent.prototype, arguments[i]);
-  }
-
-  return ReactARTComponent;
-}
-
-/**
- * Insert `node` into `parentNode` after `referenceNode`.
- */
-function injectAfter(parentNode, referenceNode, node) {
-  let beforeNode;
-  if (node.parentNode === parentNode &&
-      node.previousSibling === referenceNode) {
-    return;
-  }
-  if (referenceNode == null) {
-    // node is supposed to be first.
-    beforeNode = parentNode.firstChild;
   } else {
-    // node is supposed to be after referenceNode.
-    beforeNode = referenceNode.nextSibling;
-  }
-  if (beforeNode && beforeNode.previousSibling !== node) {
-    // Cases where `node === beforeNode` should get filtered out by earlier
-    // checks and the behavior isn't well-defined.
-    invariant(
-      node !== beforeNode,
-      'ReactART: Can not insert node before itself'
-    );
-    node.injectBefore(beforeNode);
-  } else if (node.parentNode !== parentNode) {
-    node.inject(parentNode);
+    return '';
   }
 }
 
-// ContainerMixin for components that can hold ART nodes
+/** COMPONENTS */
 
-const ContainerMixin = assign({}, ReactMultiChild, {
+class Node {
+  constructor(type, instance, props) {
+    this.type = type;
+    this.instance = instance;
 
-  /**
-   * Moves a child component to the supplied index.
-   *
-   * @param {ReactComponent} child Component to move.
-   * @param {number} toIndex Destination index of the element.
-   * @protected
-   */
-  moveChild: function(child, afterNode, toIndex, lastIndex) {
-    const childNode = child._mountImage;
-    injectAfter(this.node, afterNode, childNode);
-  },
+    this._listeners = {};
+    this._subscriptions = {};
```

Comment:
We don't need to store subscriptions because we can just rely on garbage collection to do the clean up for us. In fact, `_removeEventListeners` isn't even called.

I suppose the current ART API requires us to store the subscriptions to unsubscribe when you swap them out during updates though. Maybe we can just store them as an expando property on the ART instance as a hack?

In the normal case we don't actually use many of these listeners anyway. We attach top level listeners at the root and dispatch it ourselves.